localstack/utils/collections.py

```diff
@@ -122,14 +122,22 @@ def get_page(
 
         start_idx = 0
 
-        try:
-            start_item = next(item for item in result_list if token_generator(item) == next_token)
-            start_idx = result_list.index(start_item)
-        except StopIteration:
-            pass
+        if token_generator:
+            try:
+                start_item = next(
+                    item for item in result_list if token_generator(item) == next_token
+                )
+                start_idx = result_list.index(start_item)
+            except StopIteration:
+                pass
+        else:
+            start_idx = next_token or 0
 
         if start_idx + page_size < len(result_list):
-            next_token = token_generator(result_list[start_idx + page_size])
+            if token_generator:
+                next_token = token_generator(result_list[start_idx + page_size])
+            else:
+                next_token = start_idx + page_size
         else:
             next_token = None
```

Comment:
Imho, a token based generator is more resilient towards intermediate list modifications (while still not completely resistent).
- Using the index: Basically _every single element added or removed_ before the current page will cause shifting issues. Specific items in the list will be processed multiple times, which is quite dangerous.
- Using the token: The pagination can break if the exact element used as a token is removed. If any other elements are removed, or elements are added, it's not sure that they are going to be contained in the list.

Imho, the shifting can cause way more damage than a token-based approach.